### PR TITLE
Use relative path to get rules and decoders files in the API

### PR DESCRIPTION
**Pull request by mhamra**
|Related issue|
|---|
|[#15994](https://github.com/wazuh/wazuh/issues/15994)|

## Description
Use a relative path to get rules and decoders with the relative_dirname parameter. A format function in the spec.yml file validates this parameter.

### Implementation details of the functions get_[decoder|rule]_file
The functions get_[decoder|rule]\_file wraps get_[decoder|rule]\_files for RBAC validation and to search files. RBAC validation is done by filename only. If the relative path is passed in the filename parameter, it is NOT USED for RBAC validation.

- When the filename has no relative_dirname
  - The function get_[decoder|rule]_files is called with filename = filename and relative_path = None
- When the filename has the relative_dirname 
  - The function get_[decoder|rule]_files is called with filename = filename and relative_path = relative_dirname

```
if no results are found
  - raise an error
else if more than one result is fond
  - select the file with smaller relative_path 
else
  - select the file
 
return file content or result depending if raw is true or false
```

### CODE REVIEW

**Review by RamosFe — APPROVED** [review]
All tests passed! LGTM 🥇 

> Passed to `on-hold` because the checks don't pass.

**Review by Selutario — CHANGES_REQUESTED** [review]

File: framework/wazuh/decoder.py [hunk 1]
```diff
@@ -179,47 +179,76 @@ def get_decoders_files(status: str = None, relative_dirname: str = None, filenam
     return result
 
 
-def get_decoder_file(filename: str, raw: bool = False) -> Union[str, AffectedItemsWazuhResult]:
+def get_decoder_file(filename: str, raw: bool = False, default_ruleset: bool = True) -> Union[str, AffectedItemsWazuhResult]:
```
[the remark this answers is not in the document]
Comment (mhamra) [reply framework/wazuh/decoder.py]: Fixed in recent commit

**Comment by mhamra** [comment]
### UPDATE

git rebase 4.6.0 done

File: framework/wazuh/decoder.py [hunk 2]
```diff
@@ -196,30 +199,42 @@ def get_decoder_file(filename: str, raw: bool = False) -> Union[str, AffectedIte
...
-                                                                        f" {str(e)}"))
-        except OSError:
-            result.add_failed_item(id_=filename,
-                                   error=WazuhError(1502, extra_message=join('WAZUH_HOME', decoder_path, filename)))
 
+    # if the filename doesn't have a relative path, the search is only by name
+    # relative_dirname parameter is set to None.
+    rel_dir = relative_dirname if relative_dirname else ''
+    decoders = get_decoders_files(filename=[filename], 
+                                  relative_dirname=relative_dirname).affected_items
```
[the remark this answers is not in the document]
Comment (mhamra) [reply framework/wazuh/decoder.py:207]: - Changed in recent commit
https://github.com/wazuh/wazuh/blob/f217ffa4e64c5cbe731fd8d46619b914304e4abe/framework/wazuh/decoder.py#L203-L207
- Added new test case; see line 141
https://github.com/wazuh/wazuh/blob/f217ffa4e64c5cbe731fd8d46619b914304e4abe/framework/wazuh/tests/test_decoder.py#L135-L146

File: api/api/controllers/decoder_controller.py [hunk 3]
```diff
@@ -218,6 +219,8 @@ async def get_file(request, pretty: bool = False, wait_for_complete: bool = Fals
         Filename to download.
     raw : bool
         Whether to return the file content in raw or JSON format.
+     relative_dirname : str
+        relative directory where the decoder is located. Default None.
```
[the remark this answers is not in the document]
Comment (mhamra) [reply api/api/controllers/decoder_controller.py]: Fixed in the last pushed commit. Thanks, @Selutario

File: framework/wazuh/decoder.py [hunk 4]
```diff
@@ -179,15 +179,18 @@ def get_decoders_files(status: str = None, relative_dirname: str = None, filenam
...
     Parameters
     ----------
-    filename : str
-        Name of the decoder file.
+    filename : list. Mandatory.
+        List of one element with the complete relative path of the decoder file.
     raw : bool
         Whether to return the content in raw format (str->XML) or JSON.
+    relative_direname : str
+        relative directory where de decoder is found. Default None.
```
[the remark this answers is not in the document]
Comment (mhamra) [reply framework/wazuh/decoder.py]: Fixed in the last pushed commit. Thanks, @Selutario

File: framework/wazuh/decoder.py [hunk 5]
```diff
@@ -2,7 +2,7 @@
 # Created by Wazuh, Inc. <info@wazuh.com>.
 # This program is free software; you can redistribute it and/or modify it under the terms of GPLv2
 from os import remove
-from os.path import join, exists
+from os.path import join, exists, commonpath, basename, relpath, normpath
```
[the remark this answers is not in the document]
Comment (mhamra) [reply framework/wazuh/decoder.py]: Fixed in the last pushed commit. Thanks, @Selutario!

File: framework/wazuh/rule.py [hunk 6]
```diff
@@ -3,7 +3,7 @@
 # This program is free software; you can redistribute it and/or modify it under the terms of GPLv2
 
 from os import remove
-from os.path import exists, join
+from os.path import exists, join, abspath, commonpath, basename, relpath, normpath
```
[the remark this answers is not in the document]
Comment (mhamra) [reply framework/wazuh/rule.py]: Fixed in the last pushed commit. Thanks, @Selutario!

File: framework/wazuh/rule.py [hunk 7]
```diff
@@ -324,30 +327,42 @@ def get_rule_file(filename: str = None, raw: bool = False) -> Union[str, Affecte
...
-                                                                        f" {str(e)}"))
-        except OSError:
-            result.add_failed_item(id_=filename,
-                                   error=WazuhError(1414, extra_message=join('WAZUH_HOME', rules_path, filename)))
 
+    # if the filename doesn't have a relative path, the search is only by name
+    # relative_dirname parameter is set to None.
+    rel_dir = relative_dirname if relative_dirname else ''
+    rules = get_rules_files(filename=[filename], 
+                                  relative_dirname=relative_dirname).affected_items
```
[the remark this answers is not in the document]
Comment (mhamra) [reply framework/wazuh/rule.py]: - Fixed in the last pushed commit, but I implemented the fix differently.
https://github.com/wazuh/wazuh/blob/f217ffa4e64c5cbe731fd8d46619b914304e4abe/framework/wazuh/rule.py#L331-L335
- Added new test case; see line 242
https://github.com/wazuh/wazuh/blob/f217ffa4e64c5cbe731fd8d46619b914304e4abe/framework/wazuh/tests/test_rule.py#L237-L246

Thanks for the observation @Selutario!

File: framework/wazuh/rule.py [hunk 8]
```diff
@@ -324,30 +327,42 @@ def get_rule_file(filename: str = None, raw: bool = False) -> Union[str, Affecte
...
+    if len(rules) == 0:
+        result.add_failed_item(id_=filename, 
+                               error=WazuhError(1415, extra_message=f"{filename}"))
+        return result
+    elif len(rules) > 1:
+        # if many files match the filename criteria, 
+        # filter decoders that starts with rel_dir of the file
+        # and from the result, select the decoder with the shorter
+        # relative path length
```
[the remark this answers is not in the document]
Comment (mhamra) [reply framework/wazuh/rule.py]: Fixed in the last pushed commit. Thanks, @Selutario!

**Review by Selutario — APPROVED** [review]

**Pull request merged**